### PR TITLE
[BEAM-12445] Moving streaming inserts with new client

### PR DESCRIPTION
**Pull request by pabloem**
**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by pabloem** [comment]
Run Python 3.8 PostCommit

**Comment by pabloem** [comment]
Run Python 3.8 PostCommit

**Comment by pabloem** [comment]
Run Python PreCommit

**Comment by pabloem** [comment]
r: @chamikaramj 
cc: @ihji 

This change switches the streaming inserts client to the newer google.cloud.bigquery library. On two quick benchmark runs, I was able to double the throughput by making this change.
A shortcoming of this change is that we lose some metadata being passed in the API requests.

**Comment by pabloem** [comment]
TODO before merging: Add metadata to requests by adding it to the python-bigquery l;ibrary.

**Comment by pabloem** [comment]
Run PythonDocker PreCommit

**Comment by pabloem** [comment]
Run Portable_Python PreCommit

**Comment by pabloem** [comment]
I was able to double throughput with this change on the standard json library. Throughput should increase further for those able to use orjson

**Comment by pabloem** [comment]
@chamikaramj @ihji PTAL

File: sdks/python/apache_beam/io/gcp/bigquery.py [hunk 1]
```diff
@@ -1312,10 +1312,11 @@ def _flush_batch(self, destination):
           skip_invalid_rows=True)
       self.batch_latency_metric.update((time.time() - start) * 1000)
 
-      failed_rows = [rows[entry.index] for entry in errors]
+      failed_rows = [rows[entry['index']] for entry in errors]
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/bigquery.py:1315]: Is this due to the API change ?

File: sdks/python/apache_beam/io/gcp/bigquery_test.py [hunk 2]
```diff
@@ -809,15 +808,14 @@ def test_dofn_client_process_performs_batching(self):
     fn.process(('project_id:dataset_id.table_id', {'month': 1}))
 
     # InsertRows not called as batch size is not hit yet
-    self.assertFalse(client.tabledata.InsertAll.called)
+    self.assertFalse(client.insert_rows_json.called)
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/bigquery_test.py:811]: Do you know if the API support other formats (for example, Avro) that are more efficient ?

File: sdks/python/apache_beam/io/gcp/bigquery_test.py [hunk 3]
```diff
@@ -950,12 +943,10 @@ def store_callback(arg):
         with open(file_name_2, 'w') as f:
           json.dump(json_output, f)
 
-      res = mock.Mock()
-      res.insertErrors = []
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/bigquery_test.py:954]: Why the change to return value ?

File: sdks/python/apache_beam/io/gcp/bigquery_tools.py [hunk 4]
```diff
@@ -632,22 +639,21 @@ def _insert_all_rows(
...
+      errors = self.gcp_bq_client.insert_rows_json(
+          table_ref, json_rows=rows, row_ids=insert_ids, skip_invalid_rows=True)
+      if not errors:
         service_call_metric.call('ok')
-      for insert_error in response.insertErrors:
-        for error in insert_error.errors:
-          service_call_metric.call(error.reason)
+      for insert_error in errors:
+        service_call_metric.call(insert_error['errors'][0])
     except HttpError as e:
```
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcp/bigquery_tools.py:651]: You probably have to rebase here.

**Review by chamikaramj — COMMENTED** [review]
Thanks. Please also add a Jira.

File: sdks/python/apache_beam/io/gcp/bigquery_test.py (hunk 2 above)
Comment (pabloem) [reply sdks/python/apache_beam/io/gcp/bigquery_test.py:811]: no. The API formats everything into JSON. Other API calls use automatic JSON conversions.

File: sdks/python/apache_beam/io/gcp/bigquery.py (hunk 1 above)
Comment (pabloem) [reply sdks/python/apache_beam/io/gcp/bigquery.py:1315]: that's right.

File: sdks/python/apache_beam/io/gcp/bigquery_test.py (hunk 3 above)
Comment (pabloem) [reply sdks/python/apache_beam/io/gcp/bigquery_test.py:954]: because `insert_rows_json` returns a list of errors (empty when no errors), and `client.tabledata.InsertAll` returns an object that contains the list of errors within it

**Comment by pabloem** [comment]
Run Python 3.8 PostCommit

**Comment by pabloem** [comment]
@chamikaramj PTAL

**Comment by pabloem** [comment]
PTAL

File: sdks/python/apache_beam/io/gcp/bigquery_tools.py [hunk 5]
```diff
@@ -1114,29 +1112,16 @@ def insert_rows(
...
-      json_row = self._convert_to_json_row(row)
-      insert_id = str(self.unique_row_id) if not insert_ids else insert_ids[i]
-      final_rows.append(
-          bigquery.TableDataInsertAllRequest.RowsValueListEntry(
-              insertId=insert_id, json=json_row))
+    insert_ids = [
+        str(self.unique_row_id) if not insert_ids else insert_ids[i] for i,
+        _ in enumerate(rows)
+    ]
+    rows = [json.loads(json.dumps(r, default=default_encoder)) for r in rows]
```
Comment (ihji) [inline sdks/python/apache_beam/io/gcp/bigquery_tools.py:1119]: any reason to dump and load? sanity check?

**Review by ihji — COMMENTED** [review]
LGTM

File: sdks/python/apache_beam/io/gcp/bigquery_tools.py (hunk 4 above)
Comment (chamikaramj) [reply sdks/python/apache_beam/io/gcp/bigquery_tools.py:643]: Not rebased ? (https://github.com/apache/beam/commit/e959d621c1553a556eea8ed0b6bc9e27bd0452ac#diff-5427a5d3887eb695cefde082c58575a2372972996b547d55961abeb4f7bc3deb)

**Review by chamikaramj — COMMENTED** [review]
Thanks. LGTM.

File: sdks/python/apache_beam/io/gcp/bigquery_tools.py (hunk 5 above)
Comment (pabloem) [reply sdks/python/apache_beam/io/gcp/bigquery_tools.py:1119]: we need to pass a dictionary where its values are already json-encoded. What this line does is "encode-everything-into-json" and then "decode JSON into a dictionary with JSON-encoded fields"

File: sdks/python/apache_beam/io/gcp/bigquery_tools.py (hunk 4 above)
Comment (pabloem) [reply sdks/python/apache_beam/io/gcp/bigquery_tools.py:651]: The `raise` is there now. That's the only difference right?

File: sdks/python/apache_beam/io/gcp/bigquery_tools.py (hunk 5 above)
Comment (ihji) [reply sdks/python/apache_beam/io/gcp/bigquery_tools.py]: Okay. I thought `lambda x: loads(dumps(x))` is idempotent but looks like it's not the case here.

**Comment by pabloem** [comment]
Run Python 3.8 PostCommit

**Comment by pabloem** [comment]
Run Python PreCommit

**Comment by pabloem** [comment]
Run Python PreCommit

**Comment by quentin-sommer** [comment]
Thank you so much for this ❤️  I was about to write a ticket mentioning the poor performance of the internal client

**Comment by quentin-sommer** [comment]
Thank you for this! I was about to write a ticket mentioning the poor performance of the internal bq client

**Comment by pabloem** [comment]
hoping this will improve performance significantly. We may still need to make improvements, but we'll see : )

File: sdks/python/apache_beam/io/gcp/bigquery_tools.py [hunk 6]
```diff
@@ -292,6 +301,7 @@ def __init__(self, client=None, temp_dataset_id=None):
         http=get_new_http(),
         credentials=auth.get_service_credentials(),
         response_encoding='utf8')
+    self.gcp_bq_client = client or gcp_bigquery.Client()
```
Comment (aaltay) [inline sdks/python/apache_beam/io/gcp/bigquery_tools.py:304]: Is this a backward incompatible change, because users might pass a different type of a client that does not support `insert_rows_json` ?

**Comment by aaltay** [comment]
It seems like orjson dependency has some issue when installed on Windows. Maybe consider making it optional. (See: https://stackoverflow.com/questions/69305399/apache-beam2-32-0-installation-fails)
